### DISCUSSION TITLE
Variable substitution and configmapgenerator

### DISCUSSION THREAD

**Discussion by cciobanu98** [fluxcd/flux2#3611] [Q&A]
Hi! 

I am using the variable substitution feature from fluxCD and config map generator from kustomize.  If I use variables to create my config map , this means that config map will be the same every time, and generated hash for the config map will be the same every time. 

From my understanding this it's happening because variable substitution it's running as a post-build after kustomize build it's applied, so from kustomize perspective content of the file was not changed, but the post build changes after it. 

So, my question it's: It's there a solution for this problem to use together variable substitution and config map generator and to have a different hash when variables are updated?

**Comment by kingdonb** [comment]
The general solution is here:

https://fluxcd.io/flux/guides/helmreleases/#refer-to-values-in-configmaps-generated-with-kustomize

If this is not what you're looking for, could you be a bit more specific about what does not work in this? There are two more specific examples with different scenarios, depending on whether the values are secret or not, and whether they should be encrypted, whether they pass through Kustomize, etc.

https://fluxcd.io/flux/guides/helmreleases/#refer-to-values-in-configmap-and-secret-resources
https://fluxcd.io/flux/guides/helmreleases/#refer-to-values-in-secret-generated-with-kustomize-and-sops

The configmap generator in Kustomize runs before Flux does further post-processing. The envsubst is a post-process thing, so by the time the configmap has been generated, it is already hashed, and it is too late for envsubst to affect the content that gets hashed. You can use envsubst to fill out a configmap though, and then use `secretFrom` or `valueFrom` to adopt the value from that configmap or secret in the cluster, so that it happens as a pre-render rather than post-render substitution. Maybe this helps?

There are some clunky things in here for sure. This is a pain point that gets addressed quite heavily with templating capabilities in the Weaveworks "Weave GitOps" enterprise product, (you can do substitutions in template mode and they can take advantage of quoting functions at pre-parsing time, like you would do in a Helm chart.)

It might help if you posted an example repo, so we can suggest a specific change that will make your example work better. It is hard to pitch a general solution for everybody, when different people use different repo structures based on their organizational needs!
